### ISSUE TITLE
[HDRP] sometimes underwater postprocess is enabled above water

### ISSUE THREAD

**Issue by banan5555** [wave-harmonic/crest#655]
**Describe the bug**
Sometimes underwater postprocess is enabled above water

**Screenshots / video**

![Screenshot](https://cdn.discordapp.com/attachments/690150941751902270/756146904047288360/unknown.png)

https://www.youtube.com/watch?v=AcR_2XzU-qk

**Log**
No crest related warnings / errors.

**Versions**
Crest 4.5, Unity 2019.4.10f1, HDRP 7.4.3

**To Reproduce**
This is very rare case it shows up once a few days. When it happen once,(until restart) it is easy to reproduce like in video. But after game restart it works fine another 20 times.

**Platform**
Windows 10 x64

**Hardware**
GPU GTX 1070

**Comment by daleeidd** [comment]
Thank you for reporting. What is the value of the _Horizontal Safety Margin Multiplier_? Are you using dynamic resolution?

**Comment by banan5555** [comment]
Horizontal Safety Margin Multiplier: 0.3 (we have to use it to avoid horizontal line issue)
We don't use dynamic resolution

**Comment by daleeidd** [comment]
When the issue happens, could you check the following in the Frame Debugger? In the command list, navigate to the following: _HDRenderPipeline::Render / PostProcessing / CustomPostProcessBefore / \<unknown scope> / Draw Procedural_. Then see if __FULL_SCREEN_EFFECT_ is in Keywords.

**Comment by banan5555** [comment]
Ok, i reproduce it on editor it is related to Time provider. I run game 10 times and 9 times it work properly and last time I joined server and time was synchronize with server issue appears. I compere logs from those 10 logins and first 9 tries I join 'young' servers so time send to me was ~200-500 seconds and it works without issues, but then I join to really old one and:
'Time from server received [ms]: 84851,98'
When I stop synchronization and just provide this time to Custom time provider issue is reproduced :)
![image](https://user-images.githubusercontent.com/63231118/93685989-9d196a80-fab3-11ea-8deb-0111d89d9fba.png)

**Comment by banan5555** [comment]
screen with answer for your question here:
![image](https://user-images.githubusercontent.com/63231118/93686151-98a18180-fab4-11ea-804b-feef1bb08081.png)

**Comment by daleeidd** [comment]
Thanks. That should be perfect for testing.

In the _UnderwaterPostProcessUtils_ file, after the following line:
```csharp
bool forceFullShader = (cameraYPosition + nearPlaneFrustumWorldHeight + maxOceanVerticalDisplacement) <= seaLevel;
```
Can you add:
```csharp
forceFullShader = forceFullShader && waterHeightLevelDifference < 0f;
```
And after the following line:
```csharp
float seaLevel = OceanRenderer.Instance.SeaLevel;
```
Can you add:
```csharp
float waterHeightLevelDifference = 0f;
```
And then remove the `float` for the following line:
```csharp
float waterHeightLevelDifference = seaLevelHeightDifference;
```

**Comment by banan5555** [comment]
![image](https://user-images.githubusercontent.com/63231118/93799152-0aa5d200-fc3f-11ea-82ee-0aa0a318d45f.png)
![image](https://user-images.githubusercontent.com/63231118/93799196-1c877500-fc3f-11ea-81c7-be8c423b9fc6.png)
I looks like it is harder to reproduce and in first impression i thought that it is fixed. But when most of camera is underwater it shows again.

**Comment by daleeidd** [comment]
On the following line:
```csharp
forceFullShader = forceFullShader && waterHeightLevelDifference < 0f;
```
If you change the 0 to something like -2 then it should go away. I don't know if it will solve it for every case as I recall there being edge cases which are difficult to solve. The more you decrease that number the further down the full screen effect takes over. -2 seems like a safe and reliable value. I'll have to look into something better if I can reproduce the issue myself.

**Comment by banan5555** [comment]
Great! Thanks! No issue with '-2' :)

**Comment by huwb** [comment]
Closing as it sounds fixed(ish?), and is hopefully fixed even more with 4.10's improvement which should make the effect more stable.

**Comment by daleeidd** [comment]
The fix described here will be out for everyone in 4.12.
